### ISSUE TITLE
clang++-19: Incorrect warning `unused-private-field` even though member is used in constructor.

### ISSUE THREAD

**Issue by greg7mdp** [llvm/llvm-project#97365]
### test program: `Wunused-private-field.cpp`

```c++
// compiling this with `clang++-19 -Wall -c Wunused-private-field.cpp`
//
// produces a warning (even though `c` is used in the lambda)
//   `Wunused-private-field.cpp:19:13: warning: private field 'c' is not used [-Wunused-private-field]`
// ----------------------------------------------------------------------------------------------------
#include <functional>

struct node;

struct cluster {
   void ping();
};

void register_cb(std::function<void()>);

struct node {
   node(cluster&);
private:
   cluster &c;
};

node::node(cluster& c) : c(c) {
   register_cb([&]() { c.ping(); });
   c.ping();
}
```

### compiler output:

```
~/test ❯ clang++-19 -Wall -c Wunused-private-field.cpp                                                                                                                                     
Wunused-private-field.cpp:19:13: warning: private field 'c' is not used [-Wunused-private-field]
   19 |    cluster &c;
      |             ^
1 warning generated.
```

### compiler version

```
~/test ❯ clang++-19 --version                                                                                                                                                              
Ubuntu clang version 19.0.0 (++20240630031615+ac84ada9a169-1~exp1~20240630151739.1777)
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/lib/llvm-19/bin
```

**Comment by 5chmidti** [comment]
I don't think this is a false-positive because you are not using the field in the body, but the parameter. Reduced:

```c++
struct S {
    S(int& c) : c(c) { ++c; }

   private:
    int& c;
};
```

```
TranslationUnitDecl
`-CXXRecordDecl <line:1:1, line:6:1> line:1:8 struct S definition
  |-DefinitionData pass_in_registers trivially_copyable has_user_declared_ctor can_const_default_init
  | |-DefaultConstructor
  | |-CopyConstructor simple trivial has_const_param needs_implicit implicit_has_const_param
  | |-MoveConstructor exists simple trivial needs_implicit
  | |-CopyAssignment trivial has_const_param needs_implicit implicit_has_const_param
  | |-MoveAssignment exists trivial needs_implicit
  | `-Destructor simple irrelevant trivial needs_implicit
  |-CXXRecordDecl <col:1, col:8> col:8 implicit referenced struct S
  |-CXXConstructorDecl <line:2:5, col:29> col:5 S 'void (int &)' implicit-inline
  | |-ParmVarDecl <col:7, col:12> col:12 used c 'int &'
  | |-CXXCtorInitializer Field 0xde64f38 'c' 'int &'
  | | `-DeclRefExpr <col:19> 'int' lvalue ParmVar 0xde64d28 'c' 'int &'
  | `-CompoundStmt <col:22, col:29>
  |   `-UnaryOperator <col:24, col:26> 'int' lvalue prefix '++'
  |     `-DeclRefExpr <col:26> 'int' lvalue ParmVar 0xde64d28 'c' 'int &'     # references the parameter, not the field
  |-AccessSpecDecl <line:4:4, col:11> col:4 private
  `-FieldDecl <line:5:5, col:10> col:10 c 'int &'
```

**Comment by greg7mdp** [comment]
What? Why should I get a warning that a class member is not use, when it clearly is used. Also the warning is not on the constructor, but where the class member is declared.

**Comment by dwblaikie** [comment]
> What? Why should I get a warning that a class member is not used, when it clearly is used. Also the warning is not on the constructor, but where the class member is declared.

Because the member isn't used - the parameter is used. the name `c` in the ctor is referencing the parameter, not the member - so the member is unused.
